service/raft/group0_state_id_handler.cc

```diff
@@ -32,4 +44,77 @@ future<> group0_state_id_handler::advertise_state_id(utils::UUID state_id) {
     return _gossiper.add_local_application_state(gms::application_state::GROUP0_TOMBSTONE_GC_TIME, gms::versioned_value::gc_clock(gc_time));
 }
 
+void group0_state_id_handler::on_configuration_change(raft::server_address_set add, raft::server_address_set del) {
```

Comment:
So I am confused. I probably missed some review iterations, but I remember we discussed that we will have a simple function that runs on a timer and goes over all nodes in the current config and calculates min timestamp while looking up values in the gossiper state. Why do we need all the complicated logic of being subscribed to the raft config changes and gossiper notifications and try to maintain a coherent local state?

Reply:
To explain how we got here:
* initially the idea was to use the topology coordinator and a timer to get through the current state
* next was to create a separate timer loop in the handler itself (similar to e.g. the `cache_hitrate_calculator`) - but that also involved additional "service" being initialized from `main()`
* so at the end the handling of the state_id is being done when a change from the gossiper comes (on one side, the frequency might be higher, so the code skips if there is no change, but on the other side it allows to react to a change nearly immediately - although it is not that important in this case)
* initially the code used the endpoints retrieved from gossip to keep track about the nodes, but for the group0 the current (group0) configuration is more relevant, thus the configuration is used just to keep track of the current nodes (instead of using the gossip endpoints list)

Reply:
I remember the initial loop and I remember that @tgrabiec had doubts about it since it iterated over gossiper states and not group0 config. IMO it was enough to change it to iterated over group0 config. Current approach allows to react to changes fast (not nearly immediately since gossiper has a propagation delay and, I was assured, does not advertise new values immediately), but faster reaction to changes was never a requirement.

Reply:
See the discussion in here: https://github.com/scylladb/scylladb/pull/20394#discussion_r1783100560 ... at one point there was the code the just read the configuration from inside, but later changed to react on the configuration directly (cc @kbr-scylla )

Reply:
Well, because github PRs suck and should not be used for development ever all I can see is a comment, but not the code it was made about, but from the comment my guess is that at this point there was already subscription to gossiper changes and you accessed group0 config from there and @kbr-scylla had a concern about being in the middle of the config change that has to be handled somehow since we may not get new notification from the gossiper for a long time. Moving the logic into the timer would not have such problem.

Reply:
Another problem was that the handler was storing an IP-based map. I wanted it to be a host-ID-based map due to out pending migration to host IDs. But probably the best approach would be one which doesn't require storing a map at all. The less state we keep the better. Is such approach possible?

Reply:
> Another problem was that the handler was storing an IP-based map.

Which handler?

> I wanted it to be a host-ID-based map due to out pending migration to host IDs. But probably the best approach would be one which doesn't require storing a map at all. The less state we keep the better. Is such approach possible?

Is the timer approach I described fits the bill? The lookup in the gossiper can be done by host id (I guess).